server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDecidersTests.java

```diff
@@ -20,255 +20,137 @@
 import org.elasticsearch.cluster.routing.ShardRouting;
 import org.elasticsearch.cluster.routing.UnassignedInfo;
 import org.elasticsearch.cluster.routing.allocation.RoutingAllocation;
+import org.elasticsearch.common.transport.TransportAddress;
 import org.elasticsearch.index.Index;
 import org.elasticsearch.index.shard.ShardId;
 import org.elasticsearch.test.ESTestCase;
-import org.hamcrest.Matcher;
-import org.hamcrest.Matchers;
 
-import java.util.Collection;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+import java.util.stream.Collector;
 
 import static org.hamcrest.Matchers.equalTo;
 
 public class AllocationDecidersTests extends ESTestCase {
 
-    public void testDebugMode() {
-        verifyDebugMode(RoutingAllocation.DebugMode.ON, Matchers.hasSize(1));
+    public void testCheckAllDecidersBeforeReturningYes() {
+        var expectedDecision = new Decision.Multi();
+        var allDecisions = generateDecision(() -> Decision.YES);
+
+        verifyDecidersCall(RoutingAllocation.DebugMode.OFF, allDecisions, allDecisions.size(), expectedDecision);
     }
 
-    public void testNoDebugMode() {
-        verifyDebugMode(RoutingAllocation.DebugMode.OFF, Matchers.empty());
+    public void testCheckAllDecidersBeforeReturningThrottle() {
+        var expectedDecision = new Decision.Multi().add(Decision.THROTTLE);
+        var allDecisions = generateDecision(Decision.THROTTLE, () -> Decision.YES);
+
+        verifyDecidersCall(RoutingAllocation.DebugMode.OFF, allDecisions, allDecisions.size(), expectedDecision);
     }
 
-    public void testDebugExcludeYesMode() {
-        verifyDebugMode(RoutingAllocation.DebugMode.EXCLUDE_YES_DECISIONS, Matchers.empty());
+    public void testExitsAfterFirstNoDecision() {
+        var expectedDecision = Decision.NO;
```

Comment:
This loses a little generality from the previous tests because we sometimes used a `NO` decision decorated with an explanation. Would be good to keep that IMO.

Reply:
I am not sure I follow this, even if there is a decision with explanation only a plain NO is returned:

https://github.com/elastic/elasticsearch/blob/bb3addbd706c84e435bf66699544a2d94f810fda/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AllocationDeciders.java#L43-L59

Reply:
Right, but we want to see that we return a plain `NO` even if the decider returns a `NO` with extra info.

Reply:
Updated while separating decision that caused termination and one that is actually returned.